tests/test_applets/batchProcessing/test_batchProcessingGui.py

```diff
@@ -0,0 +1,203 @@
+from unittest import mock
+
+import pytest
+from ilastik.applets.batchProcessing.batchProcessingGui import (
+    BatchProcessingDataConstraintException,
+    BatchProcessingGui,
+    BatchRoleWidget,
+    FileListWidget,
+)
+from PyQt5.QtCore import Qt, QUrl
+
+
+def construct_widget(qtbot, widget_class, *args, **kwargs):
+    widget = widget_class(*args, **kwargs)
+    qtbot.addWidget(widget)
+    widget.show()
+    qtbot.waitForWindowShown(widget)
+    return widget
+
+
+@pytest.fixture
+def file_list_widget(qtbot):
+    return construct_widget(qtbot, FileListWidget)
+
+
+@pytest.fixture
+def batch_role_widget(qtbot):
+    return construct_widget(qtbot, BatchRoleWidget, "TestRoleName")
+
+
+@pytest.fixture()
+def role_names():
+    return ["testRoleName1", "testRoleName2", "testRoleName3"]
+
+
+@pytest.fixture()
+def parent_applet(role_names):
+    parent_applet = mock.Mock()
+    parent_applet.dataSelectionApplet.topLevelOperator.DatasetRoles.value = role_names
+    parent_applet.run_export = mock.Mock()
+
+    return parent_applet
+
+
+@pytest.fixture()
+def batch_processing_gui(qtbot, parent_applet):
+    return construct_widget(qtbot, BatchProcessingGui, parent_applet)
+
+
+@pytest.fixture
+def filenames():
+    return ["file1", "file2"]
+
+
+@pytest.fixture
+def drag_n_drop_event(filenames):
+    dndevent = mock.Mock()
+    urls = [QUrl(x) for x in filenames]
+    for url in urls:
+        url.isLocalFile = mock.Mock(return_value=True)
+    dndevent.mimeData.return_value.has_urls.return_value = True
+    dndevent.mimeData.return_value.urls.return_value = urls
+    return dndevent
+
+
+def test_FileListWidget_drag_n_drop(file_list_widget, drag_n_drop_event, filenames):
+    assert file_list_widget.count() == 0
+
+    file_list_widget.dragEnterEvent(drag_n_drop_event)
+    assert drag_n_drop_event.acceptProposedAction.called_once()
+    assert file_list_widget.count() == 0
+
+    file_list_widget.dropEvent(drag_n_drop_event)
+    assert file_list_widget.count() == len(filenames)
+    for url in filenames:
+        items = file_list_widget.findItems(url, Qt.MatchFixedString)
+        assert len(items) == 1
+
+
+def test_BatchRoleWidget(qtbot, batch_role_widget, filenames):
+    assert batch_role_widget.select_button.text() == f"Select TestRoleName Files..."
```

Comment:
I wouldn't check the contents of UI strings in tests...

Does this assert check some internal widget state? If so, then (ideally) we need a non-UI state field in widget or something similar.

Reply:
> I wouldn't check the contents of UI strings in tests...

why not? In this case I want to make sure that the gui is communicating the right string...

Reply:
This mixes GUI logic testing (how clicks and presses affect the GUI state) and GUI presentation testing (whether a certain widget has a certain color or text). If you test presentation too, then after each presentation change in code you need to change exactly the same string in a test. (If the presentation logic is indeed complex, then it probably should be moved out from presentation altogether).

Reply:
sure, I see this problem too, but in a sense I don't care if that string is set internally somewhere correctly - I really want to see that it will take the role name. As a compromise I can offer to test whether role name is in the presentation string - not the exact wording. But that has to be tested - otherwise the users will not know where to add what data.

Reply:
`assert roleName in button.text()` should be OK.